### PR TITLE
PosParser for annotation extraction from paraphrase

### PR DESCRIPTION
**Pull request by rayslxu**
A parser based on part-of-speech information to extract canonical annotations from paraphrases. 
It's implemented as a regular language using NFA, with support for the following operators:
- `*` closure
- `_` concat (implicit)
- `|` union
- `()` parentheses 
- `.` wildcard 

In addition, it uses `[]` to indicate matching group. 

This generates exactly the same annotations as the old code for the 6 schema.org domains.

### CODE REVIEW

File: lib/pos-parser/nfa.ts [hunk 1]
```diff
@@ -43,27 +47,113 @@ class State {
...
+            if (a[i + j] !== b[j])
+                match = false;
+        }
+        if (match)
+            count += 1;
+    }
+    return count;
+}
+
 export class NFA {
```
Comment (gcampax) [inline lib/pos-parser/nfa.ts:112]: At this point, NFA is not just a generic NFA engine, it's specifically for your canonical form parser. Maybe split this into two classes, one that just implements the NFA and another that has the preprocessing and POS logic?

File: test/unit/test_infix_to_postfix.js [hunk 2]
```diff
@@ -22,12 +22,19 @@ import assert from 'assert';
...
     ['( a * | b ) c', 'a * b | c _'],
     ['a b *', 'a b * _'],
-    ['a . *', 'a . * _']
+
+    // wild card
+    ['a . *', 'a . * _'],
+
+    // capturing group
+    ['a [ b ]', 'a [ b ] _'],
+    ['a [ . * c ]', 'a [ . * _ c ] _']
```
Comment (gcampax) [inline test/unit/test_infix_to_postfix.js]: Using [ ] for capturing groups is weird given the normal semantics of regular expressions...

File: test/unit/test_pos_nfa.js [hunk 3]
```diff
@@ -23,18 +23,20 @@ import { toNFA } from '../../lib/pos-parser/nfa';
 
 const TEST_CASES = [
     [
-        '( show me | find me | find | search for ) a [ $value ] $domain', ['restaurant', 'diner'], 'chinese',
+        '( show me | find me | find | search for ) ( a | ε ) [ $value ] $domain', ['restaurant', 'diner'], 'chinese',
```
Comment (gcampax) [inline test/unit/test_pos_nfa.js]: Maybe you should add the `?` operator as syntactic sugar in your regex DSL

File: lib/pos-parser/nfa.ts [hunk 4]
```diff
@@ -84,6 +84,31 @@ function arrayMatchCount(a : string[], b : string[]) : number {
...
+
+    if (current.pos === 'IN' && current.token === 'with')
+        return false;
+
+    if (current.pos.startsWith('V')) {
+        if (['has', 'have', 'had'].includes(current.token))
+            return next && ['VBN', 'VBG'].includes(next.pos!);
+        if (['is', 'are', 'was', 'were'].includes(current.token))
+            return false;
+    }
```
Comment (gcampax) [inline lib/pos-parser/nfa.ts:107]: I wonder if you should instead have multiple templates match the sentence, and use a scoring function to determine which wins...

**Review by gcampax — COMMENTED** [review]
I am honestly impressed that you implemented an NFA-based regular expression engine! I would have used backtracking and be done with it...
The code is very nice, clean and readable, which quite something for a non-trivial algorithm.

I have a few comments, but overall, let's merge this!

File: test/unit/test_pos_nfa.js (hunk 3 above)
Comment (rayslxu) [reply test/unit/test_pos_nfa.js]: > I am honestly impressed that you implemented an NFA-based regular expression engine! I would have used backtracking and be done with it...
> The code is very nice, clean and readable, which quite something for a non-trivial algorithm.
> 
> I have a few comments, but overall, let's merge this!

Thanks! The comments are all good - but let's merge this and deal with them later (once tests pass)

**Pull request merged**